Add access token support

shawnhankim opened the pull request:
**Issue:**
- https://github.com/nginxinc/nginx-openid-connect/issues/69

**Description:**
- Captured `access_token` in the NJS code.
- Stored `access_token` in the key/value store as same as the NGINX RP stores `id_token` and `refresh_token`.

shawnhankim commented:
@route443 :
- Thanks for your review in detail for the [PR](https://github.com/nginxinc/nginx-openid-connect/pull/62).
- This PR is to simplify from the previous [PR](https://github.com/nginxinc/nginx-openid-connect/pull/68).
- For you to easily manage this repo to reduce any concerns from the enhancements based on the reviews on the [PR](https://github.com/nginxinc/nginx-openid-connect/pull/62), I have divided a big PR into small PRs, and this is one of PRs.
- Let me know if you have any questions.

route443 reviewed (changes requested):
Hi Shawn. Thank you for your contribution to our nginx oidc project.
As I said in previous PRs, we are not using Conventional Commits for this project. So I wouldn't use this approach unnecessarily. Please change the commit name.

shawnhankim commented on README.md:
```diff
@@ -12,10 +12,10 @@ This repository describes how to enable OpenID Connect integration for [NGINX Pl
-  * The authorization code flow is in use
-  * NGINX Plus is configured as a relying party
-  * The IdP knows NGINX Plus as a confidential client or a public client using PKCE
+- The identity provider (IdP) supports OpenID Connect 1.0
```
Yes it is automatically changed by Markdown Formatter. Addressed this with a [separate PR](https://github.com/nginxinc/nginx-openid-connect/pull/79).

shawnhankim commented on README.md:
```diff
@@ -82,66 +84,72 @@ $ docker run -d -p 8010:8010 -v $PWD:/etc/nginx/conf.d nginx-plus nginx -g 'daem
-keyval_zone zone=refresh_tokens:1M state=conf.d/refresh_tokens.json timeout=8h;
-keyval_zone zone=oidc_pkce:128K timeout=90s;
+keyval_zone zone=oidc_id_tokens:1M     state=conf.d/oidc_id_tokens.json     timeout=1h;
+keyval_zone zone=oidc_access_tokens:1M state=conf.d/oidc_access_tokens.json timeout=1h;
```
Thanks for the great comments. Addressed it. Please let me know if I missed anything, or if you need more updates.

shawnhankim commented on frontend.conf:
```diff
@@ -27,6 +27,7 @@ server {
         # Successfully authenticated users are proxied to the backend,
         # with 'sub' claim passed as HTTP header
         proxy_set_header username $jwt_claim_sub;
+        #proxy_set_header Authorization "Bearer $access_token"; # Enable when using access token
```
Revised it

shawnhankim commented:
> Hi Shawn. Thank you for your contribution to our nginx oidc project. As I said in previous PRs, we are not using Conventional Commits for this project. So I wouldn't use this approach unnecessarily. Please change the commit name.

Hi Ivan! Thank you so much for your time to review the PR in detail. I have addressed your comments as well as the new commit name. Please let me know if there is anything need to be changed. Happy New Year!

shawnhankim commented on README.md:
````diff
@@ -223,6 +243,14 @@ proxy_set_header Host <IdP hostname>;
+  * **Invalid access token**
+    * The looping can now happen if the access token is not valid for some reason such that the setup of ID/access tokens expiration is different in the IdP.
+    * In case of this, the backend/upstream can respond with HTTP 401 "Invalid token", and the frontend can restart the flow of OpenID Connect to get new ID token or refresh tokens by calling the following directives.
+```nginx
````
Removed it.

route443 commented on README.md:
````diff
@@ -223,6 +246,12 @@ proxy_set_header Host <IdP hostname>;
 ```
 
+  * **Invalid access token**
+    Users may receive the response with HTTP 401 "invalid token" despite of successful authentication. There are several reasons why an OIDC access token might be not accepted by upstream although if it is not expired as follows:
````
```suggestion
    Users may receive a `401` response with an optional "Invalid token" message despite successful authentication. There are several reasons why an OIDC access token might not be accepted by the upstream server, even if it has not expired:
```

route443 commented on README.md:
```diff
@@ -223,6 +246,12 @@ proxy_set_header Host <IdP hostname>;
 
+  * **Invalid access token**
+    Users may receive the response with HTTP 401 "invalid token" despite of successful authentication. There are several reasons why an OIDC access token might be not accepted by upstream although if it is not expired as follows:
+    * Incorrect backend server configuration. NGINX Plus sends bearer token in the HTTP Authorization header, but the backend application expects it in a specific cookie.
```
```suggestion
    * Incorrect backend server configuration. NGINX Plus sends the bearer token in the HTTP Authorization header, but the backend application expects it in a specific cookie.
```

shawnhankim replied:
Thanks! Revised!

route443 commented on README.md:
```diff
@@ -223,6 +246,12 @@ proxy_set_header Host <IdP hostname>;
+    Users may receive the response with HTTP 401 "invalid token" despite of successful authentication. There are several reasons why an OIDC access token might be not accepted by upstream although if it is not expired as follows:
+    * Incorrect backend server configuration. NGINX Plus sends bearer token in the HTTP Authorization header, but the backend application expects it in a specific cookie.
+    * The token has been tampered with. OIDC access tokens are digitally signed by the authorization server to ensure their authenticity. If the token has been modified in any way, the signature will no longer be valid, and the token will be considered invalid.
+    * It is important to note that the scope of an OIDC access token is independent of its validity. Even if an OIDC access token is not expired and has not been revoked, it may still be considered invalid if it does not have the necessary scope for the requested action. Please check "$oidc_scopes" variable in the "openid_connect_configuration.conf" file.
```
```suggestion
    > **Note:** The scope of an OIDC access token is independent of its validity. Even if an OIDC access token is not expired and has not been revoked, it may still be considered invalid if it does not have the necessary scope for the requested action. Please check the `$oidc_scopes` variable in the `openid_connect_configuration.conf` file.
```

shawnhankim replied:
Thanks! Revised!

shawnhankim reviewed (commented):
Thanks @route443 for your review. Revised them!
